js/widgets.js

```diff
@@ -113,70 +178,149 @@ function handleIntegration (event) {
  * Main function retrieving the HTML code from the API.
  * Keep the chunk > 12 otherwise territories pages will issue more than one query.
  */
-function embedDatasets (territories, datasets) {
-  chunk(territories.concat(datasets), 12).forEach((elements) => {
+function embedDatasets (territories, datasets, dataTerritoryIdAttr, dataDatasetIdAttr) {
+  const items = territories.concat(datasets)
+  const chunkBy = 12
+  const promises = chunk(items, chunkBy).map((elements) => {
     const references = elements.map((el) => {
-      if (el.hasAttribute(TERRITORY_ID)) {
-        return `territory-${el.dataset.udataTerritoryId}`
-      } else if (el.hasAttribute(DATASET_ID)) {
-        return `dataset-${el.dataset.udataDatasetId}`
+      if (el.hasAttribute(dataTerritoryIdAttr)) {
+        return `territory-${el.dataset[camelCaseData(dataTerritoryIdAttr)]}`
+      } else if (el.hasAttribute(dataDatasetIdAttr)) {
+        return `dataset-${el.dataset[camelCaseData(dataDatasetIdAttr)]}`
       }
     })
     const url = `${baseURL}/api/1/oembeds/?references=${references}`
     // Warning: if you are tempted to use generators instead of chaining
     // promises, you'll have to use babel-polyfill which adds 300 Kb
     // once the file is converted to ES5.
-    fetch(url)
-      .then(checkStatus)
-      .then((response) => response.json())
+    return fetchJSON(url)
       .then((jsonResponse) => {
         // We match the returned list with the list of elements.
-        zip([elements, jsonResponse, references])
-          .forEach(([element, response, id]) => {
+        zip([elements, jsonResponse])
+          .forEach(([element, response]) => {
             element.innerHTML = response.html
             const integrateElement = element.querySelector('.integrate')
-            integrateElement.addEventListener('click', handleIntegration)
+            integrateElement.addEventListener('click', toggleIntegration)
           })
+        return elements
       })
       .catch(console.error.bind(console))
   })
+  Promise.all(promises)
+    .then((chunks) => {
+      // Flatten the array of datasets arrays.
+      const datasets = [].concat(...chunks)
+      if (datasets.length === items.length) {
```

Comment:
Can you have this not equal? I think you could if one network request fail. To avoid doing this, you can just remove line 207. The fetch promise will stay in the rejected state and you will never enter this then. And you'll get the console logging by line 229.